1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,565 @@
+will work exactly as pipelines currently do in PowerShell,
+so there is no breakage to the existing grammar.
+
+#### Line continuation
+
+After a pipeline chain operator,
+any newlines will be skipped in anticipation of the following pipeline.
+
+For example, the following would be a single pipeline chain:
+
+```powershell
+cmd1 &&
+    cmd2 ||
+    cmd3
+```
+
+If the end of file is reached after a pipeline chain operator,
+incomplete input will be reported so that integrating tools
+will know to keep prompting for input.
+
+#### Left-associativity
+
+Like in POSIX shells, pipeline chain operators will be left associative,
+meaning chain operators will group from left to right.
+
+For example, given the following pipeline chain:
+
+```powershell
+cmd1 && cmd2 || cmd3
+```
+
+This will be grouped as:
+
+```none
+[[cmd1 && cmd2] || cmd3]
+```
+
+As a syntax tree, this would look like:
+
+```none
+            "||"
+          /      \
+        "&&"     "cmd3"
+      /      \
+   "cmd1"    "cmd2"
+```
+
+With the syntax tree deepening on the left as more operators are chained.
+
+Semantically, this means that `cmd1 && cmd2` would be evaluated first,
+and its result used to govern the evaluation of `|| cmd3`.
+
+#### Higher precedence than `&` and `;`
+
+Pipeline chain operators will have higher precedence than
+pipeline background operators (`&`)
+or statement separators (`;`).
+
+For example:
+
+```none
+cmd1 && cmd2 &
+```
+
+Will bind as:
+
+```none
+[[cmd1 && cmd2] &]
+```
+
+Having the syntax tree:
+
+```none
+            <statement>
+           /           \
+         "&&"          "&"
+        /    \
+     "cmd1"  "cmd2"
+```
+
+The consequence of this will be that an entire pipeline chain
+can be sent to a background job for evaluation,
+rather than individual pipelines within it.
+
+### Semantics
+
+#### Pipeline "success"
+
+The `&&` and `||` operators proceed based on the "success"
+of the previous pipeline.
+
+The marker of command success proposed is the `$?` automatic variable.
+This is proposed because:
+
+- It builds off an existing PowerShell concept
+- It applies to both native commands and cmdlets/functions
+- Diverging from this would create inconsistency
+- Changes to the behaviour of `$?` would be reasonably expected to change here
+
+That is, there are the following direct equivalents:
+
+```none
+cmd1 && cmd2
+
+    |
+    v
+
+$(cmd1; if ($?) { cmd2 })
+```
+
+```none
+cmd1 || cmd2
+
+    |
+    v
+
+$(cmd1; if (-not $?) { cmd2 })
+```
+
+```none
+cmd1 && cmd2 || cmd3
+
+    |
+    v
+
+$(cmd1; if ($?) { cmd2 }; if (-not $?) { cmd3 })
+# Note that cmd1 failing runs cmd3
+```
+
+Also see the [alternate proposals section](#alternate-proposals-and-considerations).
+
+#### Pipeline output
+
+The output of a pipeline chain is the concatenation of all its pipelines,
+in the sequence of their output:
+
+```powershell
+$x = 'a','b','c' | Write-Output && Write-Output 'd'
+$x # 'a','b','c','d'
+```
+

```

Comment:
Here's an important question:

When multiple objects are returned from a command, you can capture the output from that command in different variables, like this:

```powershell
$x, $y = 1..10
```

If you run that, `$x` will have a value of 1, and `$y` will have values 2 to 10.

In pipeline chaining, if you assign the results of a pipeline chain with two pipelines to two variables like this, will the first variable receive the first object returned from the first pipeline chain that returns something, and the second variable will receive everything else? Or will the first variable receive the results from the first pipeline in the chain, and the second variable will receive the results of all other pipelines in the chain (an array of arrays, or just a single array if there is just one additional pipeline)?

For scripting purposes if it would be more useful if concatenation of pipeline data were done as an array of arrays so that you could assign a single pipeline chain to multiple variables, with each variable storing the results of one of the pipelines that were being chained. Chaining is often used for completely different commands, so we really wouldn't want to just group their output all together as a single collection of data.

If you want to skip the output of one of the pipelines in the chain, just assign that to `$null`.

For example:

```powershell
$p1,$null,$p3 = 'a','b','c' | Write-Output && Write-Output 'd' && Write-Output 'e'
$p1 # 'a','b','c'
# 'd' gets discarded since it is the result of the second pipeline in the chain and that is assigned to $null
$p3 # 'e'
```

You would need additional tests specifically for this scenario as well.

Also, in the examples above and below this you should point out that the value of `$x` will be `@('a','b','c'),@('d')`.

Reply:
In fact, PowerShell permits assignment from *any* statement. So this will work with the same semantics:

```powershell
$x, $y = if ($true) { 1; if ($true) { 2 } }
$x # 1
$y # 2
```

It will just assign each element in the pipeline to variables on the LHS until the last, which gets the tail.

But I'll mention it in the RFC for good measure.

Reply:
That wasn't my point. The point I was trying to make is that if I'm chaining pipelines, it would be of very questionable value for users to mix the results of the commands in that chain, which would often be completely different objects. At the same time, it is likely for users to want to capture the results of individual pipelines in the chain, in a more simple syntax. With this proposal in its current state we're chaining pipelines, not statements, so I can't do this:

```powershell
$r1 = p1Command && p2Command && $r3 = p3Command
```

Instead, with pipeline chaining, my point was that this:

```powershell
$r1,$null,$r3 = p1Command && p2Command && p3Command
```

would result in `$r1` containing all of the results of p1Command, the results of `$r2` would be thrown away, and `$r3` would contain all of the results of p3Command. That behaviour is new/different, but based on the proposal if we're chaining pipelines instead of statements, I really think it's the right thing to do. Otherwise, when you have no idea how many results come back from different commands, how would you capture them in different variables in pipeline chaining?

Make sense?

Reply:
You could do this, though, right?

```powershell
($r1 = p1Command) && p2Command && ($r3 = p3Command)
```

Reply:
Yes. But with that syntax I would go back to my argument that these should work more like bash, and allow that without having to use brackets.

Reply:
And even if that syntax is a possibility, I still think it makes more sense for results from individual segments in the chain to be grouped together, such that the entire result is easier to segment (array of arrays/values).